library/ssl_srv.c

```diff
@@ -3357,26 +3480,95 @@ static int ssl_parse_encrypted_pms( mbedtls_ssl_context *ssl,
         return( MBEDTLS_ERR_SSL_BAD_HS_CLIENT_KEY_EXCHANGE );
     }
 
-    mbedtls_ssl_write_version( ssl->handshake->max_major_ver,
-                       ssl->handshake->max_minor_ver,
-                       ssl->conf->transport, ver );
+    /*
+     * Decrypt the premaster secret
+     */
+#if defined(MBEDTLS_SSL_ASYNC_PRIVATE_C)
+    if( ssl->conf->f_async_decrypt_start != NULL )
+    {
+        ret = ssl->conf->f_async_decrypt_start(
+            ssl->conf->p_async_connection_ctx,
+            &ssl->handshake->p_async_operation_ctx,
+            mbedtls_ssl_own_cert( ssl ),
+            p, len );
+        switch( ret )
+        {
+        case MBEDTLS_ERR_SSL_HW_ACCEL_FALLTHROUGH:
+            /* act as if f_async_decrypt_start was null */
+            break;
+        case 0:
+            return( ssl_resume_decrypt_pms( ssl,
+                                            peer_pms,
+                                            peer_pmslen,
+                                            peer_pmssize ) );
+        case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+            return( MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS );
+        default:
+            MBEDTLS_SSL_DEBUG_RET( 1, "f_async_sign", ret );
+            return( ret );
+        }
+    }
```

Comment:
This code above looks very similar to the code used in "2.3: Compute and add the signature". Consider refactoring to ease readability and maintainability.

Reply:
It has a similar structure, but there are so many differences that an auxiliary function for that would need a lot of parameters and would need to know whether to hand over to `ssl_resume_server_key_exchange( ssl, signature_len )` or `ssl_resume_decrypt_pms( ssl, peer_pms, peer_pmslen, peer_pmssize )` in the 0 case. An auxiliary function would make the code hard to read at little benefit, and I don't think an auxiliary macro would improve maintainability.

Reply:
OK. The main motivation for this would be readability, and if making maintenance easier by sharing code structure between these two places damages readability, so be it.